From c20d8fb5db90c7ccb52a2978fe96c8fb6f6986ec Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 11:40:43 -0400
Subject: [net] ethtool: Disallow ETHTOOL_SRSSH with both indir table and hash key unchanged

Message-id: <1410176446-10440-8-git-send-email-ivecera@redhat.com>
Patchwork-id: 91655
O-Subject: [RHEL7.1 PATCH 07/10] ethtool: Disallow ETHTOOL_SRSSH with both indir table and hash key unchanged
Bugzilla: 1138823
RH-Acked-by: Michal Schmidt <mschmidt@redhat.com>
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: #1138823

Upstream commit(s):
commit 61d88c6811f216de4ec26aafe24e650dc1aeb00e
Author: Ben Hutchings <ben@decadent.org.uk>
Date:   Mon May 19 01:29:42 2014 +0100

    ethtool: Disallow ETHTOOL_SRSSH with both indir table and hash key unchanged

    This would be a no-op, so there is no reason to request it.

    This also allows conversion of the current implementations of
    ethtool_ops::{get,set}_rxfh_indir to ethtool_ops::{get,set}_rxfh
    with no change other than their parameters.

    Signed-off-by: Ben Hutchings <ben@decadent.org.uk>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/ethtool.h b/include/linux/ethtool.h
index 61c0a47..56adf9c 100644
--- a/include/linux/ethtool.h
+++ b/include/linux/ethtool.h
@@ -168,8 +168,8 @@ static inline u32 ethtool_rxfh_indir_default(u32 index, u32 n_rx_rings)
  * @set_rxfh_indir: Set the contents of the RX flow hash indirection table.
  * Will not be called if @get_rxfh_indir_size returns zero.
  * @set_rxfh: Set the contents of the RX flow hash indirection table and/or
- * hash key.  Either or both arguments may be %NULL if that attribute
- * is not to be changed.
+ * hash key.  In case only the indirection table or hash key is to be
+ * changed, the other argument will be %NULL.
  * Will only be called if one or both of @get_rxfh_indir_size and
  * @get_rxfh_key_size are implemented and return non-zero.
  * Returns a negative error code or zero.
diff --git a/net/core/ethtool.c b/net/core/ethtool.c
index 9b41871..33196f0 100644
--- a/net/core/ethtool.c
+++ b/net/core/ethtool.c
@@ -791,11 +791,14 @@ static noinline_for_stack int ethtool_set_rxfh(struct net_device *dev,
   return -EFAULT;
 
  /* If either indir or hash key is valid, proceed further.
+  * It is not valid to request that both be unchanged.
   */
  if ((user_indir_size &&
       user_indir_size != ETH_RXFH_INDIR_NO_CHANGE &&
       user_indir_size != dev_indir_size) ||
-     (user_key_size && (user_key_size != dev_key_size)))
+     (user_key_size && (user_key_size != dev_key_size)) ||
+     (user_indir_size == ETH_RXFH_INDIR_NO_CHANGE &&
+      user_key_size == 0))
   return -EINVAL;
 
  if (user_indir_size != ETH_RXFH_INDIR_NO_CHANGE)
-- 
1.7.1